<a href="https://colab.research.google.com/github/valeriafigueredo-data/Mineracao_de_Texto_e_na_Web/blob/main/2.pratica_MTW_Aula_Processamento_de_texto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Segundo Trabalho da Disciplina Mineração de Texto e na Web**

**GRUPO**

*   **André Felipe Bezerra de Souza Leão (afbsl@cesar.school)**
*   **Manuela de Lacerda Bezerra Carvalho (mlbc@cesar.school)**
*   **Valéria Cristina Andrade Rodrigues de Figueredo (vcarf@cesar.school)**

#**Segmentação de sentenças**

NLTK -> Lib em python que permite vários processamentos de texto, entre eles a segmentação de sentenças.
- Para instalar a lib

```
pip install nltk
```
- É necessário baixar o ``` pickle ``` com os tokens
```
nltk.download('punkt')
```  



In [ ]:
!pip install nltk

In [ ]:
import nltk
nltk.download('punkt')
text = "This is a English sample of the lib NLTK. NLTK is a lib for NLP."
sentence_tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')
result = sentence_tokenizer.tokenize(text)
print(len(result))

for n, sentence in enumerate(result):
  print(f"{n}: {sentence}")


2
0: This is a English sample of the lib NLTK.
1: NLTK is a lib for NLP.


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:

text_pt= "Eu tenho uma gata. Essa gata dorme no meu colo! mas eu tenho outra gata que não gosta de colo"
sentence_tokenizer = nltk.data.load('tokenizers/punkt/portuguese.pickle')
result = sentence_tokenizer.tokenize(text_pt)
print(len(result))

for n, sentence in enumerate(result):
  print(f"{n}: {sentence}")


3
0: Eu tenho uma gata.
1: Essa gata dorme no meu colo!
2: mas eu tenho outra gata que não gosta de colo


# **Tokenização de texto**

In [ ]:
from nltk.tokenize import RegexpTokenizer
tokenizer = RegexpTokenizer(r'\w+')
text = "Eu gosto de café sem açúcar, mas se o café estiver forte eu ponho açúcar. Em nossa casa as pessoas costumão tomar café com bolachas"
[s.lower() for s in tokenizer.tokenize(text)]


['eu',
 'gosto',
 'de',
 'café',
 'sem',
 'açúcar',
 'mas',
 'se',
 'o',
 'café',
 'estiver',
 'forte',
 'eu',
 'ponho',
 'açúcar',
 'em',
 'nossa',
 'casa',
 'as',
 'pessoas',
 'costumão',
 'tomar',
 'café',
 'com',
 'bolachas']

# **Remoção de StopWords**

-É necessário ter uma lista de stopwords, no NLTK basta baixar essa lista:
```nltk.download('stopwords')```


In [ ]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [ ]:
from nltk.tokenize import RegexpTokenizer
from nltk.corpus import stopwords

tokenizer = RegexpTokenizer(r'\w+')
tokens_v0 = [s.lower() for s in tokenizer.tokenize(text)]
stop_words = stopwords.words('portuguese')
tokens_v1 = [token for token in tokens_v0 if not token in stop_words]

print(tokens_v0)
print(tokens_v1)



['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['gosto', 'café', 'açúcar', 'café', 'forte', 'ponho', 'açúcar', 'casa', 'pessoas', 'costumão', 'tomar', 'café', 'bolachas']


In [ ]:
stop_words.remove("sem")
print(stop_words)

['a', 'à', 'ao', 'aos', 'aquela', 'aquelas', 'aquele', 'aqueles', 'aquilo', 'as', 'às', 'até', 'com', 'como', 'da', 'das', 'de', 'dela', 'delas', 'dele', 'deles', 'depois', 'do', 'dos', 'e', 'é', 'ela', 'elas', 'ele', 'eles', 'em', 'entre', 'era', 'eram', 'éramos', 'essa', 'essas', 'esse', 'esses', 'esta', 'está', 'estamos', 'estão', 'estar', 'estas', 'estava', 'estavam', 'estávamos', 'este', 'esteja', 'estejam', 'estejamos', 'estes', 'esteve', 'estive', 'estivemos', 'estiver', 'estivera', 'estiveram', 'estivéramos', 'estiverem', 'estivermos', 'estivesse', 'estivessem', 'estivéssemos', 'estou', 'eu', 'foi', 'fomos', 'for', 'fora', 'foram', 'fôramos', 'forem', 'formos', 'fosse', 'fossem', 'fôssemos', 'fui', 'há', 'haja', 'hajam', 'hajamos', 'hão', 'havemos', 'haver', 'hei', 'houve', 'houvemos', 'houver', 'houvera', 'houverá', 'houveram', 'houvéramos', 'houverão', 'houverei', 'houverem', 'houveremos', 'houveria', 'houveriam', 'houveríamos', 'houvermos', 'houvesse', 'houvessem', 'houvésse

In [ ]:
tokenizer = RegexpTokenizer(r'\w+')
tokens_v0 = [s.lower() for s in tokenizer.tokenize(text)]
# stop_words = stopwords.words('portuguese')
tokens_v1 = [token for token in tokens_v0 if not token in stop_words]

print(tokens_v0)
print(tokens_v1)

['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['gosto', 'café', 'sem', 'açúcar', 'café', 'forte', 'ponho', 'açúcar', 'casa', 'pessoas', 'costumão', 'tomar', 'café', 'bolachas']


# **Lemmatization**

In [ ]:
nltk.download('wordnet')
from nltk.test.portuguese_en_fixt import setup_module


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [ ]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
lemma_words = [wordnet_lemmatizer.lemmatize(word) for word in tokens_v0]
print(tokens_v0)
print(lemma_words)


['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'ma', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'a', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']


# **Stemming**

In [ ]:
nltk.download('rslp')
from nltk.stem.porter import PorterStemmer
from nltk.stem import RSLPStemmer
rslp_stemmer = RSLPStemmer()
porter_stemmer = PorterStemmer()
stem_words = [porter_stemmer.stem(word) for word in tokens_v0]
stem_words2 = [rslp_stemmer.stem(word) for word in tokens_v0]

print(tokens_v0)
print(stem_words)
print(stem_words2)

['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'ma', 'se', 'o', 'café', 'estiv', 'fort', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoa', 'costumão', 'tomar', 'café', 'com', 'bolacha']
['eu', 'gost', 'de', 'café', 'sem', 'açúc', 'mas', 'se', 'o', 'café', 'estiv', 'fort', 'eu', 'ponh', 'açúc', 'em', 'noss', 'cas', 'as', 'pesso', 'costum', 'tom', 'café', 'com', 'bolach']


[nltk_data] Downloading package rslp to /root/nltk_data...
[nltk_data]   Package rslp is already up-to-date!


## Stemming Vs Lemmatization

In [ ]:
words_sample= ['amigos', 'amigas', 'amizade', 'carreira', 'carreiras', 'tempestade']
lemma_words = [wordnet_lemmatizer.lemmatize(word) for word in words_sample]
stem_words = [porter_stemmer.stem(word) for word in words_sample]

print(words_sample)
print(lemma_words)
print(stem_words)


['amigos', 'amigas', 'amizade', 'carreira', 'carreiras', 'tempestade']
['amigo', 'amigas', 'amizade', 'carreira', 'carreiras', 'tempestade']
['amigo', 'amiga', 'amizad', 'carreira', 'carreira', 'tempestad']


# Extração de características

## Part-of-speech

Em inglês:
1. CC - Coordinating conjunction
2. CD - Cardinal number
3. DT - Determiner
4. EX - Existential there
5. FW - Foreign word
6. IN - Preposition or subordinating conjunction
7. JJ - Adjective
8. JJR - Adjective, comparative
9. JJS - Adjective, superlative
10. LS - List item marker
11. MD - Modal
12. NN - Noun, singular or mass
13. NNS - Noun, plural
14. NNP - Proper noun, singular
15. NNPS - Proper noun, plural
16. PDT - Predeterminer
17. POS - Possessive ending
18. PRP - Personal pronoun
19. PRP - 	Possessive pronoun
20. RB - Adverb
21. RBR - Adverb, comparative
22. RBS - Adverb, superlative
23. RP - Particle
24. SYM - Symbol
25. TO - to
26. UH - Interjection
27. VB - Verb, base form
28. VBD - Verb, past tense
29. VBG - Verb, gerund or present participle
30. VBN - Verb, past participle
31. VBP - Verb, non-3rd person singular present
32. VBZ - Verb, 3rd person singular present
33. WDT - Wh-determiner
34. WP - Wh-pronoun
35. WP - 	Possessive wh-pronoun
36. WRB - Wh-adverb

In [ ]:
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [ ]:
import nltk
text ="Happiness can be found, even in the darkest of times, if one only remembers to turn on the light."
tokens = [s.lower() for s in tokenizer.tokenize(text)]
nltk.pos_tag(tokens)


[('happiness', 'NN'),
 ('can', 'MD'),
 ('be', 'VB'),
 ('found', 'VBN'),
 ('even', 'RB'),
 ('in', 'IN'),
 ('the', 'DT'),
 ('darkest', 'NN'),
 ('of', 'IN'),
 ('times', 'NNS'),
 ('if', 'IN'),
 ('one', 'CD'),
 ('only', 'RB'),
 ('remembers', 'VBZ'),
 ('to', 'TO'),
 ('turn', 'VB'),
 ('on', 'IN'),
 ('the', 'DT'),
 ('light', 'NN')]

## Extração de características TF


In [ ]:
print(tokens_v0)
print(tokens_v1)

['eu', 'gosto', 'de', 'café', 'sem', 'açúcar', 'mas', 'se', 'o', 'café', 'estiver', 'forte', 'eu', 'ponho', 'açúcar', 'em', 'nossa', 'casa', 'as', 'pessoas', 'costumão', 'tomar', 'café', 'com', 'bolachas']
['gosto', 'café', 'sem', 'açúcar', 'café', 'forte', 'ponho', 'açúcar', 'casa', 'pessoas', 'costumão', 'tomar', 'café', 'bolachas']


In [ ]:
import pandas as pd
import numpy as np

df = pd.DataFrame(data=[np.zeros(11)], columns=list(set(tokens_v1)), index=["N","TF"])
quant_token = len(list(set(tokens_v1)))

for word in tokens_v0:
  if word in df.keys():
    df[word]["N"] += 1

for token in df.keys():
  df[token]["TF"] = df[token]["N"]/quant_token

df


<ipython-input-65-97e5b416e1a8>:9: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df[word]["N"] += 1
<ipython-input-65-97e5b416e1a8>:12: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values throug

forte   pessoas      café     tomar       sem     ponho  costumão  \
N   1.000000  1.000000  3.000000  1.000000  1.000000  1.000000  1.000000   
TF  0.090909  0.090909  0.272727  0.090909  0.090909  0.090909  0.090909   

       gosto    açúcar      casa  bolachas  
N   1.000000  2.000000  1.000000  1.000000  
TF  0.090909  0.181818  0.090909  0.090909

# Prática:

1. Rode o código da célula a baixo para gerar o texto base
2. Faça uma primeira exploração dos dados
 - Do que são esses dados?
 - Que tipo de "sujeiras" tem nele?
 - Quantos dados temos?
3. Limpe os dados e os separe em sentenças
4. Pra cada sentença faça:
  - Tokenização
  - Remoção de stopwords usando a lista da lib NLTK (reflita se tem alguma palavra que você colocaria/removeria da lista de stop words, qual seria?)
5. Por fim, compare a Lemmatization com Stemming de cada frase.


In [ ]:
import nltk
nltk.download('mac_morpho')
sents = nltk.corpus.mac_morpho.sents()
text = "<p>"
for s in sents[:10]:
  text += " ".join(s)
  text += ".</p> |<p>"

print(len(text))
print(text[:1000])

[nltk_data] Downloading package mac_morpho to /root/nltk_data...
[nltk_data]   Package mac_morpho is already up-to-date!


1157
<p>Jersei atinge média de Cr$ 1,4 milhão em a venda de a Pinhal em São Paulo.</p> |<p>Programe sua viagem a a Exposição Nacional do Zebu , que começa dia 25.</p> |<p>Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas.</p> |<p>A degradação de as terras por o mau uso de os solos avança em o.</p> |<p>A desertificação tornou crítica a produtividade de 52 mil km² em a região.</p> |<p>Em o dia 15 , Dia da Conservação do Solo , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação.</p> |<p>A produção brasileira de pintos de corte totalizou , em fevereiro último , 166 milhões , volume 6,79 % superior a o registrado em fevereiro de 93 , segundo dados de a Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ).</p> |<p>A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica.</p> |<p>" A melhoria de o padrão aquisitivo resulta em maior demanda de carne de frango "

# **2. Exploração dos dados**


*   Contexto: Texto com formato de Manchetes jornalísticas.
*   As sujeiras são Tags html, problemas de concordância nominal que deixa a leitura confusa e sem clareza, a repetição de letras, tal qual "a a", palavras sem sentido para o significado da frase ( A degradação das terras "avança em o" - avança onde?, como não tem complemento, foi removido), e remoção de aspas, pois as mesmas não foram abertas nas frases que as continham.
*   Quantidade de dados antes da limpeza: 206, considerando cada palavra, caractere, Tag HTML





In [ ]:
len(tokenizer.tokenize(text))

206

# **3. Limpeza de dados e separação em sentenças**

In [ ]:
import re
text_clean = re.sub(r'\<.*?\>', '', text)
text_clean = text_clean.replace('|', '')
text_clean = text_clean.replace(' em a ', ' na ')
text_clean = text_clean.replace(' de a ', ' da ')
text_clean = text_clean.replace(' a a ', ' à ')
text_clean = text_clean.replace(' de as ', ' das ')
text_clean = text_clean.replace(' de os ', ' dos ')
text_clean = text_clean.replace('Em o', 'No')
text_clean = text_clean.replace(' a o ', ' ao ')
text_clean = text_clean.replace(' de o ', ' do ')
text_clean = text_clean.replace(' em o ', ' no ')
text_clean = text_clean.replace(' em o.', '.')
text_clean = text_clean.replace(' por o ', ' pelo ')
text_clean = text_clean.replace(' " ', ' ')
text_clean = text_clean.replace(' "', ' ')
text_clean = text_clean.replace('da Pinhal', 'de Pinhal')
text_clean = text_clean.replace('São Paulo', 'SP')

print(text)
print(text_clean)

<p>Jersei atinge média de Cr$ 1,4 milhão em a venda de a Pinhal em São Paulo.</p> |<p>Programe sua viagem a a Exposição Nacional do Zebu , que começa dia 25.</p> |<p>Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas.</p> |<p>A degradação de as terras por o mau uso de os solos avança em o.</p> |<p>A desertificação tornou crítica a produtividade de 52 mil km² em a região.</p> |<p>Em o dia 15 , Dia da Conservação do Solo , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação.</p> |<p>A produção brasileira de pintos de corte totalizou , em fevereiro último , 166 milhões , volume 6,79 % superior a o registrado em fevereiro de 93 , segundo dados de a Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ).</p> |<p>A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica.</p> |<p>" A melhoria de o padrão aquisitivo resulta em maior demanda de carne de frango ".</p>

In [ ]:
len(tokenizer.tokenize(text_clean))

167

Após a limpeza dos dados, com a remoção da sujeira - tags HTLM, correção dos erros de concordância nominal, transformação de letras repetidas como "a a" para à, remoção de palavras sem sentido para o significado da frase (  A degradação das terras "avança em o" - avança onde?, como não tem complemento, foi removido) e remoção de aspas,- os dados limpos ficaram com 167 tokens.

OBS: A palavra São Paulo estava correta, gramaticalmente falando. Entretanto, foi colocado SP para que não ocorra a remoção de "são" durante o processo de remoção de stopwords e a frase fique sem sentido.

In [ ]:
entence_tokenizer = nltk.data.load('tokenizers/punkt/portuguese.pickle')
sentences = sentence_tokenizer.tokenize(text_clean)
print(len(sentences))

for n, sentence in enumerate(sentences):
  print(f"{n}: {sentence}")

10
0: Jersei atinge média de Cr$ 1,4 milhão na venda de Pinhal em SP.
1: Programe sua viagem à Exposição Nacional do Zebu , que começa dia 25.
2: Safra recorde e disponibilidade de crédito ativam vendas de máquinas agrícolas.
3: A degradação das terras pelo mau uso dos solos avança.
4: A desertificação tornou crítica a produtividade de 52 mil km² na região.
5: No dia 15 , Dia da Conservação do Solo , o único fato a festejar pode ser a Convenção Internacional sobre Desertificação.
6: A produção brasileira de pintos de corte totalizou , em fevereiro último , 166 milhões , volume 6,79 % superior ao registrado em fevereiro de 93 , segundo dados da Associação Brasileira dos Produtores de Pinto de Corte ( Apinco ).
7: A Apinco destaca em seu boletim mensal que o setor avícola está otimista com o atual programa de estabilização econômica.
8: A melhoria do padrão aquisitivo resulta em maior demanda de carne de frango .
9: O secretário da agricultura paulista , Roberto Rodrigues , aprovou o pac

## **4.1 Tokenização**

In [ ]:
tokenized_sentences = [
    [s.lower() for s in tokenizer.tokenize(sentence)]
    for sentence in sentences
]

for n, sentence in enumerate(tokenized_sentences):
  print(f"{n}: {sentence}")

0: ['jersei', 'atinge', 'média', 'de', 'cr', '1', '4', 'milhão', 'na', 'venda', 'de', 'pinhal', 'em', 'sp']
1: ['programe', 'sua', 'viagem', 'à', 'exposição', 'nacional', 'do', 'zebu', 'que', 'começa', 'dia', '25']
2: ['safra', 'recorde', 'e', 'disponibilidade', 'de', 'crédito', 'ativam', 'vendas', 'de', 'máquinas', 'agrícolas']
3: ['a', 'degradação', 'das', 'terras', 'pelo', 'mau', 'uso', 'dos', 'solos', 'avança']
4: ['a', 'desertificação', 'tornou', 'crítica', 'a', 'produtividade', 'de', '52', 'mil', 'km²', 'na', 'região']
5: ['no', 'dia', '15', 'dia', 'da', 'conservação', 'do', 'solo', 'o', 'único', 'fato', 'a', 'festejar', 'pode', 'ser', 'a', 'convenção', 'internacional', 'sobre', 'desertificação']
6: ['a', 'produção', 'brasileira', 'de', 'pintos', 'de', 'corte', 'totalizou', 'em', 'fevereiro', 'último', '166', 'milhões', 'volume', '6', '79', 'superior', 'ao', 'registrado', 'em', 'fevereiro', 'de', '93', 'segundo', 'dados', 'da', 'associação', 'brasileira', 'dos', 'produtores', 'de

# **4.2 Remoção de stopwords**

In [ ]:
stop_words = stopwords.words('portuguese')

filtered_sentences = [
    [token for token in sentence if token not in stop_words]
    for sentence in tokenized_sentences
]

for n, sentence in enumerate(filtered_sentences):
  print(f"{n}: {sentence}")

0: ['jersei', 'atinge', 'média', 'cr', '1', '4', 'milhão', 'venda', 'pinhal', 'sp']
1: ['programe', 'viagem', 'exposição', 'nacional', 'zebu', 'começa', 'dia', '25']
2: ['safra', 'recorde', 'disponibilidade', 'crédito', 'ativam', 'vendas', 'máquinas', 'agrícolas']
3: ['degradação', 'terras', 'mau', 'uso', 'solos', 'avança']
4: ['desertificação', 'tornou', 'crítica', 'produtividade', '52', 'mil', 'km²', 'região']
5: ['dia', '15', 'dia', 'conservação', 'solo', 'único', 'fato', 'festejar', 'pode', 'convenção', 'internacional', 'sobre', 'desertificação']
6: ['produção', 'brasileira', 'pintos', 'corte', 'totalizou', 'fevereiro', 'último', '166', 'milhões', 'volume', '6', '79', 'superior', 'registrado', 'fevereiro', '93', 'segundo', 'dados', 'associação', 'brasileira', 'produtores', 'pinto', 'corte', 'apinco']
7: ['apinco', 'destaca', 'boletim', 'mensal', 'setor', 'avícola', 'otimista', 'atual', 'programa', 'estabilização', 'econômica']
8: ['melhoria', 'padrão', 'aquisitivo', 'resulta', 'mai

Foi realizada a remoção de stopwords usando a lista da lib NLTK. Portanto, foi observado que seria removida a palavra são, de São Paulo. Isso afetaria a clareza e compreensão da frase. Assim, para isso não ocorrer, essa palavra - "são" deveria ser removida da lista de stop words. Entretanto, tal problema já foi solucionado na etapa de limpeza de dados, transformando São Paulo em SP.

# **5. Lemmatization X Stemming**


In [ ]:
lemma_sentences = [
    [wordnet_lemmatizer.lemmatize(word) for word in sentence]
    for sentence in filtered_sentences
]

stem_sentences = [
    [porter_stemmer.stem(word) for word in sentence]
    for sentence in filtered_sentences
]

lists = [lemma_sentences, stem_sentences]

for lst in lists:
    print(lst)

[['jersei', 'atinge', 'média', 'cr', '1', '4', 'milhão', 'venda', 'pinhal', 'sp'], ['programe', 'viagem', 'exposição', 'nacional', 'zebu', 'começa', 'dia', '25'], ['safra', 'recorde', 'disponibilidade', 'crédito', 'ativam', 'vendas', 'máquinas', 'agrícolas'], ['degradação', 'terras', 'mau', 'uso', 'solo', 'avança'], ['desertificação', 'tornou', 'crítica', 'produtividade', '52', 'mil', 'km²', 'região'], ['dia', '15', 'dia', 'conservação', 'solo', 'único', 'fato', 'festejar', 'pode', 'convenção', 'internacional', 'sobre', 'desertificação'], ['produção', 'brasileira', 'pinto', 'corte', 'totalizou', 'fevereiro', 'último', '166', 'milhões', 'volume', '6', '79', 'superior', 'registrado', 'fevereiro', '93', 'segundo', 'dado', 'associação', 'brasileira', 'produtores', 'pinto', 'corte', 'apinco'], ['apinco', 'destaca', 'boletim', 'mensal', 'setor', 'avícola', 'otimista', 'atual', 'programa', 'estabilização', 'econômica'], ['melhoria', 'padrão', 'aquisitivo', 'resulta', 'maior', 'demanda', 'carn

# **6.Conclusão**



O  presente trabalho realizou uma breve análise exploratória para compreender os dados, a limpeza dos dados, a separação em sentenças, a tokenização, a remoção de stopwords e a comparação da Lemmatization com a Stemming de cada frase.

O texto tinha formato de Manchetes jornalísticas, com diversas sujeiras que interferiam em sua clareza. Dessa forma, foi essencial realizar a limpeza dos dados. Foram removidas as Tags html, os erros de concordância nominal que deixa a leitura confusa e sem clareza, a repetição de letras, tal qual "a a"(transformou em à), palavras sem sentido para o significado da frase (  A degradação das terras "avança em o" - avança onde?, como não tem complemento, foi removido), e remoção de aspas, pois as mesmas não foram abertas nas frases que as continham.

Os dados antes da limpeza eram 206, considerando cada palavra, caractere, Tag HTML. E após a limpeza, com o procedimento informado acima, foi reduzido para 167 tokens.

A limpeza de dados, portanto, se tornou essencial no processo de identificar e corrigir ou remover dados imprecisos, incompletos, irrelevantes ou duplicados em um conjunto de dados. Esse processo é fundamental em ciência de dados, análise de dados e machine learning, pois a qualidade dos dados impacta diretamente na eficácia das análises e dos modelos.

Durante a remoção de stopwords usando a lista da lib NLTK, seria removida a palavra "são", o que resultaria em uma frase sem sentido. Portanto, tal palavra foi transformada em SP.

Por fim, sobre a Lemmatization e Stemming de cada frase, foi observado que em ambos algoritmos ocorreram falhas. No processo de stemming deve ocorrer a redução de palavras à sua raiz ou forma base, e a remoção de prefixos e sufixos. Todavia, em algumas palavras isso não ocorreu, como em "ativam", "desertificação" e outros.

E a lemmatization reduz palavras à sua forma canônica ou lema, considerando o contexto e a parte do discurso, assim as palavras "ativam", "vendas" não foram transformadas de forma coerente.

